src/programming-assistants.rst

```diff
@@ -1,23 +1,179 @@
+
+Pythia, an AI-assisted code completion system, is another Microsoft research project that has since been deployed to Visual Studio IntelliCode :cite:`svyatkovskiy2019pythia`. Pythia not only suggests method completions but also search for relevant APIs for suggestion. Pythia also uses abstract syntax trees as input to train large-scale deep learning models and attempts to predict completions within 100 ms. To compare the perform of the Pythia model, the researchers also generated models using the frequency approach and Markov-Chain approach.
+
+The frequency baseline model was generated by training a model based on the occurrence count of methods in a corpus. It was further improved by providing additional context information concerning whether or not the code was inside an "if" statement. The Markov-Chain model was model trained to determine the next method given a sequence of prior commands for certain Python libraries. For example:
+
+.. code-block:: python
+
+  os.path.isfile -> os.remove -> ?
+
+After creating the baseline models, a long short-term memory model was created with backpropagation through time, a gradient-based neural network training algorithm, applied to the model. More implementation details can be found in the paper :cite:`svyatkovskiy2019pythia`. Comparing all models created in the study, the frequency and Markov-Chain baselines were 67% and 83% accurate in predicting the expected completion within the top 5 results. The Pythia model was able to out perfom both models with an accuracy of 92% for the same metric, giving the researchers enough evidence to publish and integrate the tool into Visual Studio.
+
+.. container:: bib-item
+
+  .. bibliography:: programming-assistants.bib
+    :filter: key == 'raychev2014code'
+
+  This paper demonstrates the use of APIs to generate code completions for holes in corpuses given a query character.
+
+.. container:: bib-item
+
+  .. bibliography:: programming-assistants.bib
+    :filter: key == 'svyatkovskiy2019pythia'
+
+  This paper demonstrates the use of abstract syntax trees to train a deep learning model for code completion suggestions.
+
+Refactoring
+===========
+
+**Refactoring** refers to the process of restructuring existing code without changing the program's behavior and is a large area of interest in programming assistant research. It is intended to improve the nonfunctional attributes of a program. Typically, code refactoring applies small changes to various areas of a program and most development environments have already integrated support for basic refactoring. Detecting problems early and refactoring source code can save developers time and effort debugging in the future.
```

Comment:
Include mention of general program transformations here too, since not all suggested transformations are refactorings.